JoinColony/colonyCDapp
# Feat: Add HMAC authentication to notifications

## Description

- Add lambda for generating the user notification HMAC. When HMAC authentication is turned on in Magicbell (it hasn't been up until now, but it will be in production), this code must be passed to Magicbell when requesting the notifications to provide an extra layer of security, so that it is not enough to just know the wallet address and our Magicbell API Key to receive a list of a user's notifications. The code is generated using our Magicbell API Secret, and the user wallet address. This is why it must be calculated inside a lambda, and not client-side.

## Testing

First ensure that you have all the correct notifications env vars in your `.env` file. By now, I'm assuming most of you have, but if you haven't worked on any notifications things before then let me know and I can tell you what you need to add.

⚠️ IMPORTANT: You need to also create a `.env` file inside the folder of the new lambda which is located at `amplify/backend/function/getUserNotificationsHMAC` and add the same MAGICBELL_API_SECRET here as you did in the `.env` file for `amplify/backend/function/createUserNotificationsData`. You will then need to restart you environment so that this environment variable is picked up in the lambda.

Now, to test:

* Just go to `http://localhost:9091/planex` with the network tab open.
* Search in your network tab for a call to magicbell to fetch notifications:
<img width="1407" alt="Screenshot 2024-10-14 at 17 50 19" src="https://github.com/user-attachments/assets/cb0f97a6-7f6e-46fd-a5af-0e2713a16edc">

* Check the request headers of this request. There should be a header called `x-magicbell-user-hmac`
<img width="921" alt="Screenshot 2024-10-14 at 17 51 16" src="https://github.com/user-attachments/assets/b9e59330-aa3d-46ab-a0f1-2710c4a00198">

If you really want to, there's another thing you can do to ensure that this is working, but it's important that you please clean up afterwards.

* Go to `https://app.magicbell.com/projects/8931/settings/security` and toggle the HMAC authentication ON.
* Comment out line 69 of `src/context/Notifications/NotificationsDataContext/NotificationsDataContextProvider.tsx` so that the `userKey` prop is not passed to the `MagicbellProvider` component.
* Refresh the page. The network calls to magicbell should now be failing.
<img width="1399" alt="Screenshot 2024-10-14 at 17 53 59" src="https://github.com/user-attachments/assets/0fc21d25-2754-4df5-86f3-ba7116f82d56">

* Go back to `https://app.magicbell.com/projects/8931/settings/security` and toggle the HMAC authentication OFF (this way it does not break Magicbell for everybody else using it to develop). I will toggle this on full-time when this PR is merged and everyone has rebased onto it.


## Diffs

**New stuff** ✨

* New lambda `getUserNotificationsHMAC`

**Changes** 🏗

* Added userKey prop to the MagicbellProvider
* Added `magicbellUserKey` to `notificationsData` model (queried using lambda)

Resolves #3051
